test/e2e_node/node_shutdown_linux_test.go

```diff
@@ -0,0 +1,202 @@
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/apis/scheduling"
+	"k8s.io/kubernetes/pkg/features"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubelettypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = framework.KubeDescribe("[NodeFeature:GracefulNodeShutdown]", func() {
+	f := framework.NewDefaultFramework("graceful-node-shutdown")
+	ginkgo.Context("Graceful node shutdown", func() {
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.FeatureGates = map[string]bool{
+				string(features.GracefulNodeShutdown): true,
+			}
+			initialConfig.ShutdownGracePeriod = metav1.Duration{Duration: 20 * time.Second}
+			initialConfig.ShutdownGracePeriodCriticalPods = metav1.Duration{Duration: 10 * time.Second}
+		})
+
+		ginkgo.It("Normal shutdown", func() {
+			nodeName := getNodeName(f)
+
+			// Define test pods
+			pods := []*v1.Pod{
+				getGracePeriodOverrideTestPod("period-120", nodeName, 120, false),
+				getGracePeriodOverrideTestPod("period-5", nodeName, 5, false),
+				getGracePeriodOverrideTestPod("period-critical-120", nodeName, 120, true),
+				getGracePeriodOverrideTestPod("period-critical-5", nodeName, 5, true),
+			}
+
+			// Create pods
+			f.PodClient().CreateBatch(pods)
+
+			list, err := f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodRunning,
+					true,
+					"pod is not ready",
+				)
+			}
+
+			// Emit PrepareForShutdown signal
+			err = emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+
+			time.Sleep(5 * time.Second)
+			// Not critical pod should be shutdown
+			list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				if kubelettypes.IsCriticalPod(&pod) {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodRunning,
+						true,
+						"critical pod should not be shutdown",
+					)
+				} else {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+						true,
+						"pod should be shutdown",
+					)
+				}
+			}
+
+			time.Sleep(6 * time.Second)
+			// All pod should be shutdown
+			list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+					true,
+					"pod should be shutdown",
+				)
+			}
+		})
+
+		ginkgo.It("Interrupt shutdown", func() {
+			err := emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+			for i := 0; ; i++ {
```

Comment:
i think this loop can be converted into a `gomega.Eventually` since the goal here is basically to wait until node status goes to NotReady. I'm thinking something similar to e.g. https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/hugepages_test.go#L233-L238